posts/multiple-inheritance-crowdsales.md

```diff
@@ -0,0 +1,281 @@
+others like `_preValidatePurchase` can be overriden to implement the
+requirements of your crowdsale, but that extra behavior should be concatenated
+with the one of the parent calling super, to preserve the validations from the
+base contract. And, some functions like `_postValidatePurchase` can be just
+added as hooks in other parts of the crowdsale's lifecycle.
+
+In addition to that, we also provide some contracts for common crowdsale
+scenarios involving
+[distribution](https://github.com/OpenZeppelin/zeppelin-solidity/tree/master/contracts/crowdsale/distribution),
+[emission](https://github.com/OpenZeppelin/zeppelin-solidity/tree/master/contracts/crowdsale/emission),
+[price](https://github.com/OpenZeppelin/zeppelin-solidity/tree/master/contracts/crowdsale/price),
+and
+[validation](https://github.com/OpenZeppelin/zeppelin-solidity/tree/master/contracts/crowdsale/validation).
+These contracts are very well tested, but as we have seen here, you should be
+extra careful when combining them. Analyze the linearization of the inheritance
+graph to get a clear view of the functions that will be called and their order,
+and always always add a full suite of automated tests for your final crowdsale
+contract to make sure that it enforces all the conditions.
+
+We can confidently go a little beyond that. By now, you should have started to
+suspect that the `PreSaleWhitelistedCrowdsale` we saw above became complicated
+because of the `||` (or) condition. Things are a lot easier when all our
+conditions are merged with `&&` (and), because on that case the order of the
+conditions doesn't alter the result.
+
+Let's say that we have a crowdsale that should only allow whitelisted
+investors to buy tokens while the sale is open and the cap has not been
+reached. In this case, the condition to check would be:
+
+```
+hasStarted() && !hasEnded() && isInWhiteList(msg.sender) && isWithinCap(msg.value)
+```
+
+Our contract would be as simple as:
+
+```
+pragma solidity ^0.4.18;
+
+import "zeppelin-solidity/contracts/crowdsale/validation/TimedCrowdsale.sol";
+import "zeppelin-solidity/contracts/crowdsale/validation/WhitelistedCrowdsale.sol";
+import "zeppelin-solidity/contracts/crowdsale/validation/CappedCrowdsale.sol";
+
+contract WhitelistedTimedCappedCrowdsale is TimedCrowdsale, WhitelistedCrowdsale, CappedCrowdsale {
+
+  function WhitelistedTimedCappedCrowdsale(
+    uint256 _rate,
+    address _wallet,
+    ERC20 _token,
+    uint256 _openingTime,
+    uint256 _closingTime,
+    uint256 _cap
+  )
+    public
+    Crowdsale(_rate, _wallet, _token)
+    TimedCrowdsale(_openingTime, _closingTime)
+    CappedCrowdsale(_cap)
+    {
+    }
+
+}
+```
+
+It doesn't matter how you order the parents, all the conditions will be
+checked always. But, take this with a grain of salt. If your parent contracts
+are not super clear, they might be hiding an `||` condition in a few
+hard-to-read code statements. And, all the conditions will be checked, but the
```

Comment:
...they might be hiding an `||` condition in a few hard-to-read code statements, meaning that all the conditions will be checked, but the ...